### ISSUE TITLE
prange with creation of jitclasses

### ISSUE THREAD

**Issue by wkerzendorf** [numba/numba#4042]
We are using numba version 0.43.1 and are using it to rewrite a complex cython/c construct (I add this here as this means we can compare speeds). The code works very well in the non-parallel version of numba (in nopython mode). See this simple mainloop https://github.com/wkerzendorf/tardis/blob/numba_test/tardis/montecarlo/montecarlo_numba/base.py#L12. But produces the wrong results using the parallel decorator (and does not run any faster - even though it should very easily parallelize). We are creating jitclass objects in the main loop and are also accessing other jitclass objects (only for reading). We only write on the generated jitclass objects.

We (@chvogl and myself) would love to have input on how to debug this problem further before we go for a minimum working example.

**Comment by stuartarchibald** [comment]
Thanks for the report. Do you have a minimum working reproducer for the problem you are seeing? i.e. can you isolate it from your code base?

**Comment by stuartarchibald** [comment]
I had a go at replicating something similar using the `Bag` class from the [numba examples](https://github.com/numba/numba/blob/master/examples/jitclass.py):
```python
import numpy as np
from numba import jitclass, njit, prange
from numba import int32, float32

spec = [
    ('value', int32),
    ('array', float32[:]),
]


@jitclass(spec)
class Bag(object):
    def __init__(self, value):
        self.value = value
        self.array = np.zeros(value, dtype=np.float32)

    @property
    def size(self):
        return self.array.size

    def increment(self, val):
        for i in range(self.size):
            self.array[i] += val
        return self.array

n = 20
m = 100

@njit(parallel=True)
def foo():
    out = np.empty((m, n))
    for i in prange(m):
        clazz = Bag(n)
        clazz.increment(i)
        out[i] = clazz.array
    return out

print(foo())
foo.parallel_diagnostics(level=4)
```
and couldn't produce anything untoward.

**Comment by stuartarchibald** [comment]
Perhaps take a look at the [`parallel_diagnostics`](http://numba.pydata.org/numba-doc/latest/user/parallel.html#diagnostics) of your parallel function, see if anything suspicious has been hoisted out of the prange loop.

**Comment by wkerzendorf** [comment]
@stuartarchibald thank you - we are currently hacking on the code and enhancing it with more features and it currently does not work. But I'll try to get it back into a workable state and do `parallel_diagnostics`. I did this yesterday and it seemed very happy (something along the lines of optimal parallelization reached). I also turned on DEBUG - but it produced a lot of output that was difficult to parse. will report back.

**Comment by wkerzendorf** [comment]
@stuartarchibald 

this is what the diagnostics says:
 ```
================================================================================
 Parallel Accelerator Optimizing:  Function montecarlo_main_loop, 
/Users/wkerzend/python/tardis/tardis/montecarlo/montecarlo_numba/base.py (15)  
================================================================================


Parallel loop listing for  Function montecarlo_main_loop, /Users/wkerzend/python/tardis/tardis/montecarlo/montecarlo_numba/base.py (15) 
-------------------------------------------------------------------------------|loop #ID
@njit(**njit_dict, parallel=True, nogil=True)                                  | 
def montecarlo_main_loop(storage_model):                                       | 
    """                                                                        | 
    This is the main loop of the MonteCarlo routine that generates packets     | 
    and sends them through the ejecta.                                         | 
                                                                               | 
    Parameters                                                                 | 
    ----------                                                                 | 
    storage_model : [type]                                                     | 
        [description]                                                          | 
    """                                                                        | 
    output_nus = np.empty_like(storage_model.output_nus)                       | 
    output_energies = np.empty_like(storage_model.output_energies)             | 
    for i in prange(storage_model.no_of_packets):------------------------------| #2
        r_packet = RPacket(storage_model.r_inner[0],                           | 
                           storage_model.packet_mus[i],                        | 
                           storage_model.packet_nus[i],                        | 
                           storage_model.packet_energies[i])                   | 
        r_packet.set_line(storage_model)                                       | 
        r_packet.compute_distances(storage_model)                              | 
        #r_packet.i = i                                                        | 
        one_packet_loop(storage_model, r_packet)                               | 
        output_nus[i] = r_packet.nu                                            | 
        if r_packet.status == REABSORBED:                                      | 
            output_energies[i] = -r_packet.energy                              | 
        elif r_packet.status == EMITTED:                                       | 
            output_energies[i] = r_packet.energy                               | 
    storage_model.output_energies[:] = output_energies[:]----------------------| #0
    storage_model.output_nus[:] = output_nus[:]--------------------------------| #1
------------------------------ After Optimisation ------------------------------
Parallel structure is already optimal.
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------

```

output_energies remains 0 after it runs in parallel - in no parallel it is filled with non-zero floats

**Comment by stuartarchibald** [comment]
Thanks for the output, what is presented seems ok. Couple more things to try:
 * Switch on the highest level of diagnostics i.e. `montecarlo_main_loop.parallel_diagnostics(level=4)` this should give a load of information about hoisting which may be causing a problem here.
 * Perhaps try a development build of Numba, they are available via:
    ```bash
    conda install -c numba/label/dev numba
    ```
    (usual caveats about development versions apply!). This build would contain the patch in https://github.com/numba/numba/pull/4014 which may be relevant here.

**Comment by wkerzendorf** [comment]
@stuartarchibald thank you for your help!! level 4 diagnostics - and I'll install the dev version and report back

```
================================================================================
 Parallel Accelerator Optimizing:  Function montecarlo_main_loop, 
/Users/wkerzend/python/tardis/tardis/montecarlo/montecarlo_numba/base.py (15)  
================================================================================


Parallel loop listing for  Function montecarlo_main_loop, /Users/wkerzend/python/tardis/tardis/montecarlo/montecarlo_numba/base.py (15) 
-------------------------------------------------------------------------------|loop #ID
@njit(**njit_dict, parallel=True, nogil=True)                                  | 
def montecarlo_main_loop(storage_model):                                       | 
    """                                                                        | 
    This is the main loop of the MonteCarlo routine that generates packets     | 
    and sends them through the ejecta.                                         | 
                                                                               | 
    Parameters                                                                 | 
    ----------                                                                 | 
    storage_model : [type]                                                     | 
        [description]                                                          | 
    """                                                                        | 
    output_nus = np.empty_like(storage_model.output_nus)                       | 
    output_energies = np.empty_like(storage_model.output_energies)             | 
    for i in prange(storage_model.no_of_packets):------------------------------| #2
        r_packet = RPacket(storage_model.r_inner[0],                           | 
                           storage_model.packet_mus[i],                        | 
                           storage_model.packet_nus[i],                        | 
                           storage_model.packet_energies[i])                   | 
        r_packet.set_line(storage_model)                                       | 
        r_packet.compute_distances(storage_model)                              | 
        #r_packet.i = i                                                        | 
        one_packet_loop(storage_model, r_packet)                               | 
        output_nus[i] = r_packet.nu                                            | 
        if r_packet.status == REABSORBED:                                      | 
            output_energies[i] = -r_packet.energy                              | 
        elif r_packet.status == EMITTED:                                       | 
            output_energies[i] = r_packet.energy                               | 
    storage_model.output_energies[:] = output_energies[:]----------------------| #0
    storage_model.output_nus[:] = output_nus[:]--------------------------------| #1
--------------------------------- Fusing loops ---------------------------------
Attempting fusion of parallel loops (combines loops with similar properties)...
----------------------------- Before Optimisation ------------------------------
--------------------------------------------------------------------------------
------------------------------ After Optimisation ------------------------------
Parallel structure is already optimal.
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
 
---------------------------Loop invariant code motion---------------------------

Instruction hoisting:
loop #2:
  Has the following hoisted:
    $38.2 = global(RPacket: <class 'numba.jitclass.base.RPacket'>)
    $38.4 = getattr(value=storage__model, attr=r_inner)
    $const38.5 = const(int, 0)
    $38.8 = getattr(value=storage__model, attr=packet_mus)
    $38.12 = getattr(value=storage__model, attr=packet_nus)
    $38.16 = getattr(value=storage__model, attr=packet_energies)
    $38.28 = global(one_packet_loop: CPUDispatcher(<function one_packet_loop at 0x12c9f2e18>))
    $38.38 = global(REABSORBED: 2)
    $142.3 = global(EMITTED: 1)
  Failed to hoist the following:
    dependency: $38.6 = static_getitem(value=$38.4, index=0, index_var=$const38.5)
    dependency: $38.10 = getitem(value=$38.8, index=$parfor__index_40.155)
    dependency: $38.14 = getitem(value=$38.12, index=$parfor__index_40.155)
    dependency: $38.18 = getitem(value=$38.16, index=$parfor__index_40.155)
    dependency: $38.21 = getattr(value=$r_packet.43, attr=set_line)
    dependency: $38.23 = call $38.21(storage__model, func=$38.21, args=[Var(storage__model, /Users/wkerzend/python/tardis/tardis/montecarlo/montecarlo_numba/base.py (26))], kws=(), vararg=None)
    dependency: $38.25 = getattr(value=$r_packet.43, attr=compute_distances)
    dependency: $38.27 = call $38.25(storage__model, func=$38.25, args=[Var(storage__model, /Users/wkerzend/python/tardis/tardis/montecarlo/montecarlo_numba/base.py (26))], kws=(), vararg=None)
    dependency: $38.31 = call $38.28(storage__model, $r_packet.43, func=$38.28, args=[Var(storage__model, /Users/wkerzend/python/tardis/tardis/montecarlo/montecarlo_numba/base.py (26)), Var($r_packet.43, /Users/wkerzend/python/tardis/tardis/montecarlo/montecarlo_numba/base.py (32))], kws=(), vararg=None)
    dependency: $38.37 = getattr(value=$r_packet.43, attr=status)
    dependency: $38.39 = $38.37 == $38.38
    dependency: $142.2 = getattr(value=$r_packet.43, attr=status)
    dependency: $142.4 = $142.2 == $142.3
--------------------------------------------------------------------------------
```

**Comment by wkerzendorf** [comment]
@stuartarchibald it seems to work with the dev version - however it is slower than without parallelization. We are currently simplifying the code and will try again after.

**Comment by stuartarchibald** [comment]
Great, thanks for checking. Might also be worth checking that it is indeed still parallelizing the loop (it should do). That it is correct is an improvement, working out what's pessimising the code is another matter :) If you manage to simplify the problem to a MWE, or make it trivial for this to be built and tested, I can probably take a look.

**Comment by wkerzendorf** [comment]
@stuartarchibald I think we are sort of done here with this issue - I'll still try to get a runnable version to you as there are some more questions.

**Comment by stuartarchibald** [comment]
@wkerzendorf thanks for getting back to us, I'll close this issue. As noted, more than happy to take a look at compact/reproducible examples, it'd be helpful if you could open a new issue for them and refer back to this one as needed. Thanks. 

> Thank you very much for your help and this fantastic tool!!

No problem, thanks for using Numba :)

PS: If you get stuck/need a hand, the Numba core developers are also available on [gitter.im](https://gitter.im/numba/numba) real time chat.
